libcontainer/container_linux.go

```diff
@@ -619,6 +619,22 @@ func (c *Container) newParentProcess(p *Process) (parentProcess, error) {
 	}
 
 	if safeExe != nil {
+		// Because we want to add safeExe to the set of ExtraFiles, if the fd of safeExe is
+		// too small, go stdlib will dup3 it to another fd, or dup3 a other fd to this fd,
+		// then it will cause the fd type cmd.Path refers to a random path. (#4294)
+		if int(safeExe.Fd()) <= stdioFdCount+len(cmd.ExtraFiles)+1 {
+			maxFd, err := utils.GetMaxFds()
+			if err != nil {
+				return nil, fmt.Errorf("unable to get the max opened fd of current process: %w", err)
+			}
+			maxFd = maxFd + 1
+			if err := unix.Dup3(int(safeExe.Fd()), maxFd, unix.F_DUPFD_CLOEXEC); err != nil {
+				return nil, fmt.Errorf("unable to dup3 the fd from %d to %d: %w", safeExe.Fd(), maxFd, err)
+			}
+			cmd.Path = "/proc/self/fd/" + strconv.Itoa(maxFd)
```

Comment:
should we do `safeExec.Close()` somewhere around here?

Reply:
In my brain, `unix.Dup3` would close the old fd silently if there is no error, I'll verify it.

Reply:
You are right, it will not close the old fd automatically.